rust/navigator/src/lib.rs

```diff
@@ -35,23 +35,23 @@ lazy_static! {
 }
 
 ///This should be called from UI; returns new UI information as JSON
-pub fn do_action(action_str: &str, details_str: &str, secret_seed_phrase: &str) -> String {
+pub fn do_action(
+    action: Action,
+    details_str: &str,
+    secret_seed_phrase: &str,
+) -> Result<ActionResult, String> {
     //If can't lock - debounce failed, ignore action
     //
     //guard is defined here to outline lifetime properly
     let guard = STATE.try_lock();
     match guard {
-        Ok(mut state) => {
-            let action = Action::parse(action_str);
-            let details = (*state).perform(action, details_str, secret_seed_phrase);
-            (*state).generate_json(&details)
-        }
+        Ok(mut state) => state.perform(action, details_str, secret_seed_phrase),
         Err(TryLockError::Poisoned(_)) => {
             //TODO: maybe more grace here?
             //Maybe just silently restart navstate? But is it safe?
             panic!("Concurrency error! Restart the app.");
         }
-        Err(TryLockError::WouldBlock) => "".to_string(),
+        Err(TryLockError::WouldBlock) => Err("WouldBlock".to_string()),
```

Comment:
Just don't output this error and make this back into unfallible. This is hardware debouncing, or, in other words, solution for concurrency involving a human. This is not an error that should be shown or logged anywhere. This is also silently rejected on native frontend side.